src/profilers/profile-serializer.ts

```diff
@@ -0,0 +1,146 @@
+ */
+export function serializeTimeProfile(
+    prof: TimeProfile, intervalMicros: number) {
+  let samples: Array<perftools.profiles.Sample> = [];
+  let locations: Array<perftools.profiles.Location> = [];
+  let functions: Array<perftools.profiles.Function> = [];
+  let locationMap: Map<number, perftools.profiles.Location> = new Map();
+  let functionMap: Map<number, perftools.profiles.Function> = new Map();
+  let strings = [''];
+
+  let sampleValueType = createSampleValueType();
+  let timeValueType = createTimeValueType();
+
+  serializeNode(prof.topDownRoot);
+
+  return {
+    sampleType: [sampleValueType, timeValueType],
+    sample: samples,
+    location: locations,
+    function: functions,
+    stringTable: strings,
+    // opt drop_frames
+    // opt keep_frames
+    timeNanos: 1000 * 1000 * prof.endTime,                         // Nanos
+    durationNanos: 1000 * 1000 * (prof.endTime - prof.startTime),  // Nanos
+
+    periodType: timeValueType,
+    period: intervalMicros
+  };
+
+  /**
+   * Adds samples from a node and it's children to the fields tracking
+   * profile serialization.
+   *
+   * @param node - the node which is serialized
+   * @param stack - the stack trace to the current node.
+   */
+  function serializeNode(root: TimeProfileNode) {
+    let entries: Entry[] = [];
+    // don't include root node in serialized profile, start with it's children.
+    for (let child of root.children) {
+      entries.push({node: child, stack: []});
+    }
+    while (entries.length > 0) {
+      let entry = entries.pop();
+      if (entry !== undefined) {
+        let node = entry.node;
+        let stack = entry.stack;
+        let location = getLocation(node);
+        stack.unshift(location.id as number);
+        if (node.hitCount > 0) {
+          const sample = new perftools.profiles.Sample({
+            locationId: stack,
+            value: [node.hitCount, node.hitCount * intervalMicros]
+          });
+          samples.push(sample);
+        }
+        for (let child of node.children) {
+          entries.push({node: child, stack: stack.slice(0)});
+        }
+      }
+    }
+  }
+
+  function getLocation(node: TimeProfileNode): perftools.profiles.Location {
+    const id = node.callUid;
+    let location = locationMap.get(id);
+    if (location !== undefined) {
+      return location;
+    }
+    location = new perftools.profiles.Location({id: id, line: [getLine(node)]});
+    locations.push(location);
+    locationMap.set(id, location);
+    return location;
+  }
+
+  function getLine(node: TimeProfileNode): perftools.profiles.Line {
+    return new perftools.profiles.Line(
+        {functionId: getFunction(node).id, line: node.lineNumber});
+  }
+
+  function getFunction(node: TimeProfileNode): perftools.profiles.Function {
+    const id = node.callUid;
+    let f = functionMap.get(id);
+    if (f !== undefined) {
+      return f;
+    }
+    const name = getIndexOrAdd(node.functionName || '(anonymous)', strings);
```

Comment:
getIndexOrAdd is O(n), which makes the profile conversion quadratic the number of functions plus the number of source files. I think we need a more efficient lookup.